IPIP/0298-denylists-for-ipfs-nodes-and-gateways.md

```diff
@@ -0,0 +1,243 @@
+
+## Motivation
+
+The current [Bad Bit Denylist](https://badbits.dwebops.pub/denylist.json), which is a list of hashed CIDs, has multiple major drawbacks for implementation. One is lacking support for content path, it's hard to block an IPNS path only based on CIDs. The other is lacking description and status code response for each anchor. The description is especially for hashed blocking items, since the hashing is unidirectional, the description can outline the reason why it's been blocked thus increasing maintainability.
+
+A well-thought-out list schema can ease the implementation of denylists, and bring consensus on how content blocking works for gateways, and node operators in the community.
+
+## Architecture
+
+```mermaid
+sequenceDiagram
+  autonumber
+  actor User
+  participant IPFS Node
+  participant List Provider
+  note over IPFS Node: Discover List <CID> by List Provider
+  IPFS Node->>List Provider: ipfs://<CID>
+  List Provider->>IPFS Node: List
+  loop For each entry
+    IPFS Node->>IPFS Node: Add entry to rule engine
+  end
+  User->>IPFS Node: ipfs://bafy...badcid
+  IPFS Node->>IPFS Node: Process bafy...badcid through rule engine
+  IPFS Node->>User: 403 Forbidden
+```
+
+### List Provider
+
+A List Provider is an entity building and providing a list of content to be allowed/denied. They can provide their lists through the IPFS network or via [an HTTP Gateway endpoint](../http-gateways/README.md).
+
+### IPFS Node
+
+An IPFS node is a service that interacts with the IPFS network. This could be because it's providing content to the network, because it's retrieving content, or because it's serving IPFS hosted content through an HTTP interface.
+
+To interact with content lists, it operates a rule engine. The rule engine decides what action should be taken for a particular request.
+For instance, if CID bafy...badcid is denied, the bitswap engine should not request it, and access to bafy...badcid through a node exposed gateway are going to 403 forbidden.
+
+### User
+
+User is an actor that can make request to the IPFS Node. It's considered authenticated already.
+Access to the node can be made via an HTTP gateway, Bitswap, or another transport method exposed by the IPFS Node.
+
+It is understood that the User might be in control of the IPFS Node, and is able to configure the rule engine as it suits them.
+
+## Schemas
+
+### Definition
+
+The schemas below are defined as [IPLD schemas](https://ipld.io/specs/schemas/).
+
+```ipldsch
+type List struct {
+  Action Int # Suggested action to be taken
+  StatusCode optional Int # HTTP status code to be returned by IPFS gateways by default
+
+  Entries optional Entry[] # Entries included in the list
+  Imports optional Bytes # CID to a chain of entries included in the list
+
+  Description optional String # Context to be shared with the list consumers
+}
+
+type Entry struct {
+  Type Int # Type of the content string. Refer to CID_V1, ABSOLUTE_PATH, PARTIAL_PATH, SHA256_CID_V1
```

Comment:
There should be a type that has has exactly the same meaning as the current badbits anchor.